cns/deviceplugin/server.go

```diff
@@ -0,0 +1,153 @@
+		deviceCheckInterval:  deviceCheckInterval,
+	}
+}
+
+// Run starts the grpc server and blocks until an error or context is cancelled. Wait on Ready to know when the server is ready.
+func (s *Server) Run(ctx context.Context) error {
+	grpcServer := grpc.NewServer()
+	v1beta1.RegisterDevicePluginServer(grpcServer, s)
+	defer close(s.listAndWatchDoneChan)
+
+	l, err := net.Listen("unix", s.address)
+	if err != nil {
+		return errors.Wrap(err, "error listening on socket")
+	}
+	errChan := make(chan error, 2) //nolint:gomnd // disabled in favor of readability
+	go func() {
+		errChan <- grpcServer.Serve(l)
+	}()
+	defer grpcServer.Stop()
+
+	select {
+	case err := <-errChan:
+		return errors.Wrap(err, "error running grpc server")
+	case <-ctx.Done():
+		return nil
+	}
+}
+
+// Ready blocks until the server is ready
+func (s *Server) Ready(ctx context.Context) error {
+	c, err := grpc.DialContext(ctx, s.address, grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithBlock(), //nolint:staticcheck // TODO: Move to grpc.NewClient method
+		grpc.WithContextDialer(func(ctx context.Context, addr string) (net.Conn, error) {
+			conn, err := (&net.Dialer{}).DialContext(ctx, "unix", addr)
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to dial context")
+			}
+			return conn, nil
+		}),
+	)
+	if err != nil {
+		return errors.Wrap(err, "error dialing local grpc server")
+	}
+	if err := c.Close(); err != nil {
+		return errors.Wrap(err, "error closing connection to local grpc server")
+	}
+	return nil
+}
+
+func (s *Server) Allocate(_ context.Context, req *v1beta1.AllocateRequest) (*v1beta1.AllocateResponse, error) {
+	s.logger.Info("allocate request", zap.Any("req", *req))
+	resps := make([]*v1beta1.ContainerAllocateResponse, len(req.ContainerRequests))
+	for i, containerReq := range req.ContainerRequests {
+		resp := &v1beta1.ContainerAllocateResponse{
+			Envs: make(map[string]string),
+		}
+		for j := range containerReq.DevicesIDs {
+			resp.Envs[fmt.Sprintf("NIC_%d", j)] = containerReq.DevicesIDs[j]
+		}
+		resps[i] = resp
+	}
+	r := &v1beta1.AllocateResponse{
+		ContainerResponses: resps,
+	}
+	return r, nil
+}
+
+func (s *Server) ListAndWatch(_ *v1beta1.Empty, stream v1beta1.DevicePlugin_ListAndWatchServer) error {
+	// send the initial count right away
+	advertisedCount := s.deviceCounter.getDeviceCount()
+	devices := make([]*v1beta1.Device, advertisedCount)
+	for i := range devices {
+		devices[i] = &v1beta1.Device{
+			ID:     fmt.Sprintf("NIC-%d", i),
+			Health: v1beta1.Healthy,
+		}
+	}
+	if err := stream.Send(&v1beta1.ListAndWatchResponse{
+		Devices: devices,
+	}); err != nil {
+		return errors.Wrap(err, "error sending listAndWatch response")
+	}
+
+	// every interval, check if the current count has changed from what we've previously sent, and if so, send the new count
+	for {
+		select {
+		case <-s.listAndWatchDoneChan:
+			return nil
+		case <-time.After(s.deviceCheckInterval):
```

Comment:
choosing not to use ticker was intentional (but maybe erroneous?), my understanding of ticker was that if one operation was going slow, the next tickers will fire immediately, so suppose you had a ticker every second, but one iteration takes 30s for some reason, you'll get the next 30 ticks as fast as you can possibly process them. And it made me nervous potentially calling `stream.Send` in that kind of case, so I liked that `time.After` starts when it is called regardless of how long the previous iteration took

Reply:
Understand why you were worried about this, but the Go authors thought of it:
> 	// Give the channel a 1-element time buffer.
	// If the client falls behind while reading, we drop ticks
	// on the floor until the client catches up.

and 
https://cs.opensource.google/go/go/+/refs/tags/go1.22.6:src/time/tick.go;l=16-17

You won't get 30 ticks in your scenario, you will get 1 tick at the next 1 second interval from when the ticker started (which could be very soon! but will definitely be within 1 second from now.) VS the `time.After` behavior of: you will get another tick _after another one second_.
The problem with `time.After` is that if there is nothing listening, the backing timer/goroutine is leaked because the tick is never consumed. We can hit that if we enter the select and `listAndWatchDoneChan` closes and we return while we are sleeping one second for `time.After`.

Reply:
makes sense, agreed let's use ticker

Reply:
Updated to use Ticker